pkg/operator/operator.go

```diff
@@ -241,76 +273,237 @@ func (k *Kontroller) withLeaderElection() error {
+// checkBeforeReboot gets all nodes with the before-reboot=true label and checks
+// if all of the configured before-reboot annotations are set to true. If they
+// are, it deletes the before-reboot=true label and sets reboot-ok=true to tell
+// the agent that it is ready to start the actual reboot process.
+// If it goes to set reboot-ok=true and finds that the node no longer wants a
+// reboot, then it just deletes the before-reboot=true label.
+// If there is an error getting the list of nodes or updating any of them, an
+// error is immediately returned.
+func (k *Kontroller) checkBeforeReboot() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed listing nodes: %v", err)
+	}
+
+	preRebootNodes := k8sutil.FilterNodesByRequirement(nodelist.Items, beforeRebootReq)
+
+	for _, n := range preRebootNodes {
+		if hasAllAnnotations(n, k.beforeRebootAnnotations) {
+			glog.V(4).Infof("Deleting label %q for %q", constants.LabelBeforeReboot, n.Name)
+			glog.V(4).Infof("Setting annotation %q to true for %q", constants.AnnotationOkToReboot, n.Name)
+			err = k8sutil.UpdateNodeRetry(k.nc, n.Name, func(node *v1api.Node) {
+				delete(node.Labels, constants.LabelBeforeReboot)
+
+				// make sure that the node still wants to reboot
+				if !wantsRebootSelector.Matches(fields.Set(node.Annotations)) {
+					glog.Warningf("Node %v no longer wanted to reboot while we were trying to mark it so: %v", node.Name, node.Annotations)
+					return
+				}
+
+				node.Annotations[constants.AnnotationOkToReboot] = constants.True
+			})
+			if err != nil {
+				return fmt.Errorf("Failed to update node %q: %v", n.Name, err)
+			}
 		}
 	}
 
-	nodelist, err = k.nc.List(v1meta.ListOptions{})
+	return nil
+}
+
+// checkAfterReboot gets all nodes with the after-reboot=true label and checks
+// if  all of the configured after-reboot annotations are set to true. If they
+// are, it deletes the after-reboot=true label and sets reboot-ok=false to tell
+// the agent that it has completed it's reboot successfully.
+// If there is an error getting the list of nodes or updating any of them, an
+// error is immediately returned.
+func (k *Kontroller) checkAfterReboot() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
 	if err != nil {
-		glog.Infof("Failed listing nodes: %v", err)
-		return
+		return fmt.Errorf("Failed listing nodes: %v", err)
+	}
+
+	postRebootNodes := k8sutil.FilterNodesByRequirement(nodelist.Items, afterRebootReq)
+
+	for _, n := range postRebootNodes {
+		if hasAllAnnotations(n, k.afterRebootAnnotations) {
+			glog.V(4).Infof("Deleting label %q for %q", constants.LabelAfterReboot, n.Name)
+			glog.V(4).Infof("Setting annotation %q to false for %q", constants.AnnotationOkToReboot, n.Name)
+			err = k8sutil.UpdateNodeRetry(k.nc, n.Name, func(node *v1api.Node) {
+				delete(node.Labels, constants.LabelAfterReboot)
```

Comment:
Similar to above, where do we wanna cleanup the after-reboot annotations? I'm missing that part